python/TestHarness/TestHarness.py

```diff
@@ -46,10 +47,23 @@ def findTestRoot(start=os.getcwd(), method=os.environ.get('METHOD', 'opt')):
 class TestHarness:
 
     @staticmethod
-    def buildAndRun(argv, app_name, moose_dir):
+    def buildAndRun(argv, app_name, moose_dir, unit=None, unit_dir=None, unit_run_make=True):
         harness = TestHarness(argv, moose_dir, app_name=app_name)
         harness.findAndRunTests()
-        sys.exit(harness.error_code)
+        unit = bool(os.getenv('MOOSE_RUN_TESTS_WITH_UNIT', False)) if unit is None else unit
+        rcode = TestHarness.buildAndRunUnit(unit_dir, unit_run_make) if unit else 0
+        sys.exit(harness.error_code or rcode)
+
+    @staticmethod
+    def buildAndRunUnit(unit_dir=None, run_make=True):
+        """Execute unit tests."""
+        unit_dir = unit_dir or os.path.join(os.getcwd(), 'unit')
+        if run_make:
+            cmd = ['make', '-C', unit_dir, '-j', str(int(os.cpu_count()/2))]
```

Comment:
Hmm - Do we really want the TestHarness in the business of building? Why not do this through the make process?

Reply:
Because I don't understand make.

Reply:
I'm fairly certain we already have all the rules for the unit test apps in the app.mk file. We just don't use them there because we have separate Makefiles for the unit test apps. It should be a few lines to get our make files to build unit apps by default. Maybe I'll take a look at that before this goes in (with all my ample spare time).